In [178]:
# The following line will import KNeighborsClassifier "Class"
# KNeighborsClassifier is name of a "sklearn class" to perform "KNN Classification" 
from pandas import DataFrame
from sklearn.neighbors import KNeighborsClassifier

# Importing the required packages and libraries
# we will need numpy and pandas later
import numpy as np
import pandas as pd

# reading a CSV file directly from Web, and store it in a pandas DataFrame:
# "read_csv" is a pandas function to read csv files from web or local device:
iris_df = pd.DataFrame

iris_df = pd.read_csv('https://raw.githubusercontent.com/mpourhoma/CS4661/master/iris.csv')

In [179]:
# checking the dataset by printing every 10 lines:
iris_df[0::10]

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sepal_length</th>
      <th>sepal_width</th>
      <th>petal_length</th>
      <th>petal_width</th>
      <th>species</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>5.1</td>
      <td>3.5</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
    </tr>
    <tr>
      <th>10</th>
      <td>5.4</td>
      <td>3.7</td>
      <td>1.5</td>
      <td>0.2</td>
      <td>setosa</td>
    </tr>
    <tr>
      <th>20</th>
      <td>5.4</td>
      <td>3.4</td>
      <td>1.7</td>
      <td>0.2</td>
      <td>setosa</td>
    </tr>
    <tr>
      <th>30</th>
      <td>4.8</td>
      <td>3.1</td>
      <td>1.6</td>
      <td>0.2</td>
      <td>setosa</td>
    </tr>
    <tr>
      <th>40</th>
      <td>5.0</td>
      <td>3.5</td>
      <td>1.3</td>
      <td>0.3</td>
      <td>setosa</td>
    </tr>
    <tr>
      <th>50</th>
      <td>7.0</td>
      <td>3.2</td>
      <td>4.7</td>
      <td>1.4</td>
      <td>versicolor</td>
    </tr>
    <tr>
      <th>60</th>
      <td>5.0</td>
      <td>2.0</td>
      <td>3.5</td>
      <td>1.0</td>
      <td>versicolor</td>
    </tr>
    <tr>
      <th>70</th>
      <td>5.9</td>
      <td>3.2</td>
      <td>4.8</td>
      <td>1.8</td>
      <td>versicolor</td>
    </tr>
    <tr>
      <th>80</th>
      <td>5.5</td>
      <td>2.4</td>
      <td>3.8</td>
      <td>1.1</td>
      <td>versicolor</td>
    </tr>
    <tr>
      <th>90</th>
      <td>5.5</td>
      <td>2.6</td>
      <td>4.4</td>
      <td>1.2</td>
      <td>versicolor</td>
    </tr>
    <tr>
      <th>100</th>
      <td>6.3</td>
      <td>3.3</td>
      <td>6.0</td>
      <td>2.5</td>
      <td>virginica</td>
    </tr>
    <tr>
      <th>110</th>
      <td>6.5</td>
      <td>3.2</td>
      <td>5.1</td>
      <td>2.0</td>
      <td>virginica</td>
    </tr>
    <tr>
      <th>120</th>
      <td>6.9</td>
      <td>3.2</td>
      <td>5.7</td>
      <td>2.3</td>
      <td>virginica</td>
    </tr>
    <tr>
      <th>130</th>
      <td>7.4</td>
      <td>2.8</td>
      <td>6.1</td>
      <td>1.9</td>
      <td>virginica</td>
    </tr>
    <tr>
      <th>140</th>
      <td>6.7</td>
      <td>3.1</td>
      <td>5.6</td>
      <td>2.4</td>
      <td>virginica</td>
    </tr>
  </tbody>
</table>
</div>

In [180]:
# Defining a function to convert "categorical" labels to "numerical" labels
# This is optional, because the latest revision of sklearn accepts non-numerical labels too!

def categorical_to_numeric(x):
    if x == 'setosa':
        return 0
    elif x == 'versicolor':
        return 1
    elif x == 'virginica':
        return 2

In [181]:
# Applying the function on species column and adding corrsponding numerical label column:

iris_df['label'] = iris_df['species'].apply(categorical_to_numeric)

# checking the dataset by printing every 10 lines:

iris_df[0::10]

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sepal_length</th>
      <th>sepal_width</th>
      <th>petal_length</th>
      <th>petal_width</th>
      <th>species</th>
      <th>label</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>5.1</td>
      <td>3.5</td>
      <td>1.4</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0</td>
    </tr>
    <tr>
      <th>10</th>
      <td>5.4</td>
      <td>3.7</td>
      <td>1.5</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0</td>
    </tr>
    <tr>
      <th>20</th>
      <td>5.4</td>
      <td>3.4</td>
      <td>1.7</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0</td>
    </tr>
    <tr>
      <th>30</th>
      <td>4.8</td>
      <td>3.1</td>
      <td>1.6</td>
      <td>0.2</td>
      <td>setosa</td>
      <td>0</td>
    </tr>
    <tr>
      <th>40</th>
      <td>5.0</td>
      <td>3.5</td>
      <td>1.3</td>
      <td>0.3</td>
      <td>setosa</td>
      <td>0</td>
    </tr>
    <tr>
      <th>50</th>
      <td>7.0</td>
      <td>3.2</td>
      <td>4.7</td>
      <td>1.4</td>
      <td>versicolor</td>
      <td>1</td>
    </tr>
    <tr>
      <th>60</th>
      <td>5.0</td>
      <td>2.0</td>
      <td>3.5</td>
      <td>1.0</td>
      <td>versicolor</td>
      <td>1</td>
    </tr>
    <tr>
      <th>70</th>
      <td>5.9</td>
      <td>3.2</td>
      <td>4.8</td>
      <td>1.8</td>
      <td>versicolor</td>
      <td>1</td>
    </tr>
    <tr>
      <th>80</th>
      <td>5.5</td>
      <td>2.4</td>
      <td>3.8</td>
      <td>1.1</td>
      <td>versicolor</td>
      <td>1</td>
    </tr>
    <tr>
      <th>90</th>
      <td>5.5</td>
      <td>2.6</td>
      <td>4.4</td>
      <td>1.2</td>
      <td>versicolor</td>
      <td>1</td>
    </tr>
    <tr>
      <th>100</th>
      <td>6.3</td>
      <td>3.3</td>
      <td>6.0</td>
      <td>2.5</td>
      <td>virginica</td>
      <td>2</td>
    </tr>
    <tr>
      <th>110</th>
      <td>6.5</td>
      <td>3.2</td>
      <td>5.1</td>
      <td>2.0</td>
      <td>virginica</td>
      <td>2</td>
    </tr>
    <tr>
      <th>120</th>
      <td>6.9</td>
      <td>3.2</td>
      <td>5.7</td>
      <td>2.3</td>
      <td>virginica</td>
      <td>2</td>
    </tr>
    <tr>
      <th>130</th>
      <td>7.4</td>
      <td>2.8</td>
      <td>6.1</td>
      <td>1.9</td>
      <td>virginica</td>
      <td>2</td>
    </tr>
    <tr>
      <th>140</th>
      <td>6.7</td>
      <td>3.1</td>
      <td>5.6</td>
      <td>2.4</td>
      <td>virginica</td>
      <td>2</td>
    </tr>
  </tbody>
</table>
</div>

In [182]:
# Creating the Feature Matrix for iris dataset:

# create a python list of feature names that would like to pick from the dataset:
feature_cols = ['sepal_length','sepal_width','petal_length','petal_width']

# use the above list to select the features from the original DataFrame
X = iris_df[feature_cols]  

# print the first 5 rows
X

<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sepal_length</th>
      <th>sepal_width</th>
      <th>petal_length</th>
      <th>petal_width</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>5.1</td>
      <td>3.5</td>
      <td>1.4</td>
      <td>0.2</td>
    </tr>
    <tr>
      <th>1</th>
      <td>4.9</td>
      <td>3.0</td>
      <td>1.4</td>
      <td>0.2</td>
    </tr>
    <tr>
      <th>2</th>
      <td>4.7</td>
      <td>3.2</td>
      <td>1.3</td>
      <td>0.2</td>
    </tr>
    <tr>
      <th>3</th>
      <td>4.6</td>
      <td>3.1</td>
      <td>1.5</td>
      <td>0.2</td>
    </tr>
    <tr>
      <th>4</th>
      <td>5.0</td>
      <td>3.6</td>
      <td>1.4</td>
      <td>0.2</td>
    </tr>
    <tr>
      <th>5</th>
      <td>5.4</td>
      <td>3.9</td>
      <td>1.7</td>
      <td>0.4</td>
    </tr>
    <tr>
      <th>6</th>
      <td>4.6</td>
      <td>3.4</td>
      <td>1.4</td>
      <td>0.3</td>
    </tr>
    <tr>
      <th>7</th>
      <td>5.0</td>
      <td>3.4</td>
      <td>1.5</td>
      <td>0.2</td>
    </tr>
    <tr>
      <th>8</th>
      <td>4.4</td>
      <td>2.9</td>
      <td>1.4</td>
      <td>0.2</td>
    </tr>
    <tr>
      <th>9</th>
      <td>4.9</td>
      <td>3.1</td>
      <td>1.5</td>
      <td>0.1</td>
    </tr>
    <tr>
      <th>10</th>
      <td>5.4</td>
      <td>3.7</td>
      <td>1.5</td>
      <td>0.2</td>
    </tr>
    <tr>
      <th>11</th>
      <td>4.8</td>
      <td>3.4</td>
      <td>1.6</td>
      <td>0.2</td>
    </tr>
    <tr>
      <th>12</th>
      <td>4.8</td>
      <td>3.0</td>
      <td>1.4</td>
      <td>0.1</td>
    </tr>
    <tr>
      <th>13</th>
      <td>4.3</td>
      <td>3.0</td>
      <td>1.1</td>
      <td>0.1</td>
    </tr>
    <tr>
      <th>14</th>
      <td>5.8</td>
      <td>4.0</td>
      <td>1.2</td>
      <td>0.2</td>
    </tr>
    <tr>
      <th>15</th>
      <td>5.7</td>
      <td>4.4</td>
      <td>1.5</td>
      <td>0.4</td>
    </tr>
    <tr>
      <th>16</th>
      <td>5.4</td>
      <td>3.9</td>
      <td>1.3</td>
      <td>0.4</td>
    </tr>
    <tr>
      <th>17</th>
      <td>5.1</td>
      <td>3.5</td>
      <td>1.4</td>
      <td>0.3</td>
    </tr>
    <tr>
      <th>18</th>
      <td>5.7</td>
      <td>3.8</td>
      <td>1.7</td>
      <td>0.3</td>
    </tr>
    <tr>
      <th>19</th>
      <td>5.1</td>
      <td>3.8</td>
      <td>1.5</td>
      <td>0.3</td>
    </tr>
    <tr>
      <th>20</th>
      <td>5.4</td>
      <td>3.4</td>
      <td>1.7</td>
      <td>0.2</td>
    </tr>
    <tr>
      <th>21</th>
      <td>5.1</td>
      <td>3.7</td>
      <td>1.5</td>
      <td>0.4</td>
    </tr>
    <tr>
      <th>22</th>
      <td>4.6</td>
      <td>3.6</td>
      <td>1.0</td>
      <td>0.2</td>
    </tr>
    <tr>
      <th>23</th>
      <td>5.1</td>
      <td>3.3</td>
      <td>1.7</td>
      <td>0.5</td>
    </tr>
    <tr>
      <th>24</th>
      <td>4.8</td>
      <td>3.4</td>
      <td>1.9</td>
      <td>0.2</td>
    </tr>
    <tr>
      <th>25</th>
      <td>5.0</td>
      <td>3.0</td>
      <td>1.6</td>
      <td>0.2</td>
    </tr>
    <tr>
      <th>26</th>
      <td>5.0</td>
      <td>3.4</td>
      <td>1.6</td>
      <td>0.4</td>
    </tr>
    <tr>
      <th>27</th>
      <td>5.2</td>
      <td>3.5</td>
      <td>1.5</td>
      <td>0.2</td>
    </tr>
    <tr>
      <th>28</th>
      <td>5.2</td>
      <td>3.4</td>
      <td>1.4</td>
      <td>0.2</td>
    </tr>
    <tr>
      <th>29</th>
      <td>4.7</td>
      <td>3.2</td>
      <td>1.6</td>
      <td>0.2</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>120</th>
   

In [183]:
# checking the size of Feature Matix X:

print(X.shape)

(150, 4)


In [184]:
# select a Series of labels (the last column) from the DataFrame
# y = iris_df['label'] # this is the index that we gave to the labels
# OR:
y = iris_df['species'] # this is the original categorical labels (the latest revision of sklearn accepts non-numerical labels)

# checking the label vector by printing every 10 values
y[::10]

0          setosa
10         setosa
20         setosa
30         setosa
40         setosa
50     versicolor
60     versicolor
70     versicolor
80     versicolor
90     versicolor
100     virginica
110     virginica
120     virginica
130     virginica
140     virginica
Name: species, dtype: object

In [185]:
k = 3
knn = KNeighborsClassifier(n_neighbors=3)


In [186]:
knn.fit(X, y)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=3, p=2,
           weights='uniform')

In [187]:
# Randomly splitting the original dataset into training set and testing set
# The function"train_test_split" from "sklearn.cross_validation" library performs random splitting.
# "test_size=0.3" means that pick 30% of data samples for testing set, and the rest (70%) for training set.

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=6)

In [188]:
# print the size of the training set:
print(X_train.shape)
print(y_train.shape)

(90, 4)
(90,)


In [189]:
# print the size of the testing set:
print(X_test.shape)
print(y_test.shape)


(60, 4)
(60,)


In [190]:
print('—x test—')
print(X_test)
print('\n')
print('—y test—')
print(y_test)

—x test—
     sepal_length  sepal_width  petal_length  petal_width
4             5.0          3.6           1.4          0.2
116           6.5          3.0           5.5          1.8
2             4.7          3.2           1.3          0.2
23            5.1          3.3           1.7          0.5
123           6.3          2.7           4.9          1.8
96            5.7          2.9           4.2          1.3
134           6.1          2.6           5.6          1.4
39            5.1          3.4           1.5          0.2
137           6.4          3.1           5.5          1.8
53            5.5          2.3           4.0          1.3
127           6.1          3.0           4.9          1.8
81            5.5          2.4           3.7          1.0
115           6.4          3.2           5.3          2.3
135           7.7          3.0           6.1          2.3
74            6.4          2.9           4.3          1.3
119           6.0          2.2           5.0          1.5
105  

In [191]:
# Training ONLY on the training set:
knn.fit(X_train, y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=3, p=2,
           weights='uniform')

In [192]:
# Testing on the testing set:

y_predict = knn.predict(X_test)
print(y_predict)

['setosa' 'virginica' 'setosa' 'setosa' 'virginica' 'versicolor'
 'virginica' 'setosa' 'virginica' 'versicolor' 'virginica' 'versicolor'
 'virginica' 'virginica' 'versicolor' 'versicolor' 'virginica'
 'versicolor' 'versicolor' 'setosa' 'setosa' 'virginica' 'setosa' 'setosa'
 'versicolor' 'versicolor' 'versicolor' 'virginica' 'setosa' 'versicolor'
 'setosa' 'versicolor' 'setosa' 'setosa' 'versicolor' 'virginica'
 'versicolor' 'virginica' 'versicolor' 'setosa' 'setosa' 'virginica'
 'versicolor' 'versicolor' 'setosa' 'setosa' 'versicolor' 'setosa'
 'setosa' 'versicolor' 'virginica' 'virginica' 'virginica' 'setosa'
 'virginica' 'setosa' 'setosa' 'setosa' 'versicolor' 'virginica']


In [193]:
# Function "accuracy_score" from "sklearn.metrics" will perform element-to-element comparision and returns the 
# percent of correct predictions:

from sklearn.metrics import accuracy_score

# Example:
y_pred = [0, 2, 1, 1]
y_actual = [0, 1, 2, 1]

score = accuracy_score(y_actual, y_pred)

print(score)


0.5


In [194]:
# We can now compare the "predicted labels" for the Testing Set with its "actual labels" to evaluate the accuracy 
# Function "accuracy_score" from "sklearn.metrics" will perform the element-to-element comparision and returns the 
# portion of correct predictions:

from sklearn.metrics import accuracy_score

accuracy = accuracy_score(y_test, y_predict)

print(accuracy)


0.9666666666666667


In [195]:
# checking results

results = pd.DataFrame()

results['actual'] = y_test 
results['prediction'] = y_predict 

print(results)

         actual  prediction
4        setosa      setosa
116   virginica   virginica
2        setosa      setosa
23       setosa      setosa
123   virginica   virginica
96   versicolor  versicolor
134   virginica   virginica
39       setosa      setosa
137   virginica   virginica
53   versicolor  versicolor
127   virginica   virginica
81   versicolor  versicolor
115   virginica   virginica
135   virginica   virginica
74   versicolor  versicolor
119   virginica  versicolor
105   virginica   virginica
51   versicolor  versicolor
92   versicolor  versicolor
32       setosa      setosa
37       setosa      setosa
120   virginica   virginica
44       setosa      setosa
0        setosa      setosa
55   versicolor  versicolor
72   versicolor  versicolor
87   versicolor  versicolor
102   virginica   virginica
30       setosa      setosa
93   versicolor  versicolor
45       setosa      setosa
59   versicolor  versicolor
16       setosa      setosa
13       setosa      setosa
133   virginica  ver

In [196]:
k_arr = [1, 5, 7, 11, 15, 27, 59]
accuracy_list = []

for arr in k_arr:
    k_loop = arr
    knn_loop = KNeighborsClassifier(n_neighbors=arr)
    knn_loop.fit(X, y)
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=6)
    # training on training set
    knn_loop.fit(X_train, y_train)
    # testing on testing set
    y_predict = knn_loop.predict(X_test)
    # Example:
    y_pred = [0, 2, 1, 1]
    y_actual = [0, 1, 2, 1]

    score = accuracy_score(y_actual, y_pred)
    accuracy = accuracy_score(y_test, y_predict)
    accuracy_list.append(accuracy)
    

for item in accuracy_list:
    print(item)


0.95
0.9833333333333333
0.9666666666666667
0.9666666666666667
0.9333333333333333
0.9166666666666666
0.8166666666666667


## Accuracy gets slightly better if k is increased slightly, but gets worse as k goes higher

In [197]:
single_feature_accuracy_list = []
two_feature_accuracy_list = []

k = 3
knn = KNeighborsClassifier(n_neighbors=k)

feature_cols = [['sepal_length'], ['sepal_width'], ['petal_length'], ['petal_width']]

for item in feature_cols:
    X = iris_df[item]
    knn.fit(X, y)
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=6)
    knn.fit(X_train, y_train)
    accuracy = accuracy_score(y_test, y_predict)
    single_feature_accuracy_list.append(accuracy)

for item in single_feature_accuracy_list:
    print(item)
    
single_feature_accuracy_list.clear()


0.8166666666666667
0.8166666666666667
0.8166666666666667
0.8166666666666667


In [198]:
k = 3
knn = KNeighborsClassifier(n_neighbors=k)

knn.fit(X, y)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=3, p=2,
           weights='uniform')

In [199]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=6)

knn.fit(X_train, y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=3, p=2,
           weights='uniform')

In [200]:
accuracy = accuracy_score(y_test, y_predict)
single_feature_accuracy_list.append(accuracy)

In [201]:
feature_cols = ['sepal_width']

X = iris_df[feature_cols]

knn.fit(X, y)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=6)

knn.fit(X_train, y_train)

y_predict = knn.predict(X_test)
accuracy = accuracy_score(y_test, y_predict)
single_feature_accuracy_list.append(accuracy)

In [202]:
feature_cols = ['petal_length']

X = iris_df[feature_cols]

knn.fit(X, y)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=6)

knn.fit(X_train, y_train)
y_predict = knn.predict(X_test)


In [203]:
accuracy = accuracy_score(y_test, y_predict)
single_feature_accuracy_list.append(accuracy)

In [204]:
feature_cols = ['petal_width']

X = iris_df[feature_cols]

knn.fit(X, y)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=6)

knn.fit(X_train, y_train)
y_predict = knn.predict(X_test)


In [205]:
accuracy = accuracy_score(y_test, y_predict)
single_feature_accuracy_list.append(accuracy)

In [206]:
for item in single_feature_accuracy_list:
    print(item)

0.8166666666666667
0.5666666666666667
0.9333333333333333
0.95


## Petal Width is the best feature (95% accuracy) followed by petal length (93.33% accuracy) as the second best feature

In [207]:
feature_pairs = [['sepal_length', 'sepal_width'], ['sepal_length', 'petal_length'], ['sepal_length', 'petal_width'],
                 ['sepal_width', 'petal_length'], ['sepal_width', 'petal_width'], ['petal_length', 'petal_width']]

for item in feature_pairs:
    X = iris_df[item]
    knn.fit(X, y)
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=6)
    knn.fit(X_train, y_train)
    y_predict = knn.predict(X_test)

    accuracy = accuracy_score(y_test, y_predict)
    two_feature_accuracy_list.append(accuracy)

for element in two_feature_accuracy_list:
    print(element)


0.8166666666666667
0.9833333333333333
0.95
0.95
0.95
0.9666666666666667


`feature_cols = ['sepal_length', 'petal_length']`

## The two best features are Sepal Length and Petal Length (98.33%)

## The best feature pair from part (f) does not contain both of the first best feature and the second best feature, but only includes one of them being Petal Length

## We cannot conclude that the “best two features” for classification are the first best feature along with the second best feature together because the best feature pair does not include both the first best and the second best from testing single feature labels